### ISSUE TITLE
transition to manifest V3

### ISSUE THREAD

**Issue by neogeographica** [neogeographica/singleplayer_scripts#26]
https://developer.chrome.com/blog/mv2-transition/

Looks like V2 will be gone next year. Will need to have a look at https://developer.chrome.com/docs/extensions/mv3/intro/mv3-migration/

**Comment by neogeographica** [comment]
One sort-of-blocking issue with this change. https://www.google.com/search?q=chrome+extension+manifest+v3+createObjectURL

**Comment by neogeographica** [comment]
Initial work is on the [manifest-v3](https://github.com/neogeographica/singleplayer_scripts/tree/manifest-v3) branch.

That createObjectURL problem is a head-scratcher.

**Comment by neogeographica** [comment]
A [data URL](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/Data_URLs) would be a way to deal with this for small files, but the base64-encoded size of the file cannot exceed 500MB. Some reports say even much smaller sizes (e.g. 64MB) can crash the browser. It also just kind of offends me to have to stick the entire file into the URL.

As a data point, the Arcane Dimensions zipfile is about 300MB, then when base64-encoded it is over 400MB.

This crbug is relevant: https://bugs.chromium.org/p/chromium/issues/detail?id=1224027

(Since it's been open for over a year without progress, I'm dubious about anything being done about it.)

**Comment by neogeographica** [comment]
For workaround #3 mentioned in the above crbug, a couple of relevant docs...

To change the URL of the active tab: https://developer.chrome.com/docs/extensions/reference/tabs/#method-update

To respond to fetch events: https://developer.mozilla.org/en-US/docs/Web/API/FetchEvent

More fetch-event example code, including a constructed response: https://googlechrome.github.io/samples/service-worker/fallback-response/

Response constructor docs: https://developer.mozilla.org/en-US/docs/Web/API/Response/Response

**Comment by neogeographica** [comment]
OK let's work on a more detailed plan...

==========

The click-listener function set by addListener eventually ends up with the content in a "chunks" array, which it creates a Blob from.

Currently it proceeds to invoke chrome.downloads.download on URL.createObjectURL(blob). That's no longer allowed.

So instead, at this point we need to get the browser to open some "virtual URL" we compose, e.g. "chrome-extension://open-with-quake/\<file>.qz", then we can intercept and handle that.

To support that kind of URL, need this in the manifest:
```
    "web_accessible_resources": [
      {
        "resources": [ "*.qz" ],
        "matches": [ "<all_urls>" ],
        "use_dynamic_url": true
      }
    ]
```

The workaround mentioned in the links above talks about using chrome.tabs.update to change the current tab, but what if the tab was closed in the meantime? We could instead use [chrome.tabs.create](https://developer.chrome.com/docs/extensions/reference/tabs/#method-create) to create some new tab (without changing the active tab), pointed at that URL. There might be an issue there though in getting our fetch-event listener to fire.

Well, ignore that for a moment. What would that fetch-listener look like? We can create it within the click-listener function to make a closure that has access to the Blob `blob` and the filename `file`.

So, something like:

```javascript
// Compose our virtual URL.
virtualUrl = chrome.runtime.getURL(file + ".qz");
// Set a (self-removing) callback to respond to it.
const callback = function handleFetch(event) {
  if (event.request.url === virtualUrl) {
    event.respondWith(
      new Response(
        blob,
        {
          status: 200,
          headers: {'Content-Disposition': 'attachment'}
        }
      )
    );
    self.removeEventListener('fetch', handleFetch);
  }
}
self.addEventListener('fetch', callback);
// Open a new tab to trigger the fetch.
let newTab = await chrome.tabs.create({
  active: false,
  url: virtualUrl
});
if (!newTab) {
  // XXX error, remove listener and bail out
}
```

OK now, at least one and perhaps two issues.
* We should delete the tab after the response is sent, but we don't know the tab identity when creating the fetch-listener. I guess it's possible this might not be necessary for a tab with an attachment content-disposition URL, will test, but I'm guessing it will be necessary.
* As mentioned above, the listener may not respond to that other tab. But it might!

To solve the first problem, we could first create the tab with a blank URL. Something like:

```javascript
// Open a new blank tab.
let newTab = await chrome.tabs.create({active: false, url: "about:blank"});
if (!newTab) {
  // XXX error, bail out
}
let tabId = tab.id;
// Compose our virtual URL.
virtualUrl = chrome.runtime.getURL(file + ".qz");
// Set a (self-removing) callback to respond to it.
const callback = function handleFetch(event) {
  if (event.request.url === virtualUrl) {
    event.respondWith(
      new Response(
        blob,
        {
          status: 200,
          headers: {'Content-Disposition': 'attachment'}
        }
      )
    );
    self.removeEventListener('fetch', handleFetch);
    chrome.tabs.remove(tabId);
  }
}
self.addEventListener('fetch', callback);
// Update the tab to trigger the fetch.
let updatedTab = await chrome.tabs.update(
  tabId,
  {active: false, url: virtualUrl}
);
if (!updatedTab) {
  // XXX error, remove listener and tab and bail out
}
```

It's possible that we could remove the tab as soon as it is updated, since that should trigger the fetch, although I don't know if that will make much observable difference. I.e. could move `chrome.tabs.remove(tabId)` out of the callback and down to the end of this code snippet.

Now, I need to play with that to figure out all the terrible problems and whether or not the fetch listener fires.

Also probably there are other error cases to handle. Like, I dunno, if the tab update reportedly succeeds but we never get the fetch event, do we need to worry about timing out eventually and cleaning up?

==========

OK the other part of this is that if we can't use chrome.downloads.download then the handleDownloadStarted callback is useless. Need to remove that and replace its functionality.

What does it actually do? Basically, just call either `setStatus(false, "")` or `setStatus(false, "err")` depending on whether the blob download succeeded.

In the new approach above, `event.respondWith` has no observable error result. It can throw exceptions in some cases, so maybe I should catch those? But as a first cut, we could just do `setStatus(false, "")` right after `event.respondWith`. The `setStatus(false, "err")` call could be used in the XXX-marked error cases.

**Comment by neogeographica** [comment]
First problem encountered:

"Event handler of 'fetch' event must be added on the initial evaluation of worker script."

Soooo OK need some substantial reworking there.

(Or I guess just some yucky global variables. Would need to do explicit blob-deletion in that case.)

**Comment by neogeographica** [comment]
Hey it works! At first glance. No need to explicitly delete the tab either.

Next problem is that I'm no longer allowed to make a DOMParser to search for `meta http-equiv="refresh"` in the returned page. But as far as this particular issue goes it's time to close it.

**Comment by neogeographica** [comment]
Eh, re-opening. I forgot this was an umbrella issue about manifest V3. :-)

So, current issue: not allowed to use a DOMParser to parse the fetched page. Sigh.

If I was interested in parsing some actual page in the web browser I could use a content script, but I think for my usecase of just needing to parse content that I have as a string in a variable, I'm out of luck.

This is not a super essential feature, it's just to deal with sites that do a redirect using an HTML element instead of a header. I think link-shortener sites were the main offender there. But maybe there's a way to get it to work, like a quick-and-dirty search for that meta element using regex (yeah I know you can't generally parse HTML with regex).

Maybe I could send the text-to-parse in a message to a content script, then send some result back in the response?

https://developer.chrome.com/docs/extensions/mv3/messaging/

Also kinda gross, but maybe less gross than regex crimes.

**Comment by neogeographica** [comment]
Related crbug: https://bugs.chromium.org/p/chromium/issues/detail?id=1056354
and workaround-in-progress-maybe ("offscreen documents"): https://bugs.chromium.org/p/chromium/issues/detail?id=1339382

It's not clear to me exactly how I would use "offscreen documents" to do HTML parsing. The proposed create function seems to want a URL of a "real" document to load, and then there's no particular way to interact with it although maybe it could do message-passing like you would do with a content script. Even in that case I dunno how you would initiate a message from the extension to the offscreen doc (other direction would be doable tho).

Wait-and-see, I guess.

It is crazy how broken V3 is for various things while Google is apparently holding fast to the January deadline. I am also having a hard time believing any of this stuff is the right solution for just wanting to parse HTML.

**Comment by neogeographica** [comment]
Looking at some of the [test code](https://chromium-review.googlesource.com/c/chromium/src/+/3723924/4/chrome/browser/extensions/offscreen_document_browsertest.cc#251) submitted for this Chrome feature, it appears that [chrome.runtime.sendMessage](https://developer.chrome.com/docs/extensions/reference/runtime/#method-sendMessage) can be used to send from the extension to the offscreen doc or vice versa (so, different from talking to content scripts).

Theroretically I think could have an HTML resource page "offscreen.html" bundled with the extension that just contains the javascript for doing the necessary parseFromString logic, as a message listener that operates on the message content and sends back the result. Maybe:
```html
<html>
<head> 
</head> 
<body> 
<script>
chrome.runtime.onMessage.addListener((msg, sender, sendResponse) => {
  // XXX here do the parsing of "msg" to get redirectUrl; set to null if no redirect
  sendResponse(redirectUrl);
}
</script>
</body> 
</html>
```
(Although I suppose to be nice I should put the JS in a separate script file.)

Then the extension JS would do something like
```javascript
let redirectUrl = await chrome.runtime.sendMessage(page);
// XXX here appropriately handle redirectUrl, including whether it is null
```

The message argument does have a size limit of 128 MB, hopefully in practice no page HTML would be that big.

Of course the extension would also have to initially call something like
```javascript
await chrome.offscreen.createDocument(
  {
    url: 'offscreen.html',
    reasons: ['DOM_SCRAPING'],
    justification: 'parsing received HTML page to find any meta-refresh redirect'
  }
);
```
and do some manifest changes (maybe just adding "offscreen" to permissions?), etc.

Cf. the [test extension](https://chromium.googlesource.com/chromium/src/+/e230ab25a9e5e05ce4dc6a945bbd79cd43b735bc/chrome/test/data/extensions/api_test/offscreen/basic_document_management).

There's something about offscreen docs getting reaped if they are "no longer active", and I don't know what that really means in this context.

I can't test this unless I switch to Canary chrome builds and use the "--enable-features=ExtensionsOffscreenDocuments" commandline arg. Or get a non-Canary build of at least 109.0.5414.0 I think.

Edit: Of course, Google apparently doesn't want me to use Canary on Linux. I guess maybe Dev will break that 109.0.5414.0 barrier soon. Check https://chromereleases.googleblog.com/2022/ in a few days to see.

**Comment by neogeographica** [comment]
Oh hey, the deadline has been moved to June.

**Comment by neogeographica** [comment]
Another thing to keep an eye on: [5 minute lifetime for service workers](https://bugs.chromium.org/p/chromium/issues/detail?id=1152255). Although maybe longer if they are not "idle"? Not sure.

Almost all downloads here would be quicker than 5 minutes though, knock on wood.

**Comment by neogeographica** [comment]
saving here a potentially useful tip about service worker debugging: https://groups.google.com/a/chromium.org/g/chromium-extensions/c/3QAinUhCiPY/m/gt8f9eASAgAJ

**Comment by joel-bluedata** [comment]
Testing with the new Chrome dev-channel version, everything generally works! Except... yeah, the service worker becomes "inactive" even while I am using it.

**Comment by neogeographica** [comment]
Testing with the new Chrome dev-channel version, everything generally works! Except... yeah, the service worker becomes "inactive" even while I am using it.

**Comment by neogeographica** [comment]
Trying out a workaround on the manifest-v3 branch, based on suggestions from [the ever-useful wOxxOm](https://stackoverflow.com/a/66618269). Basically, as long as the menu handler is busy: find an open tab and ask it to "connect" back to us every 250s, to trigger the hey-I'm-not-idle logic. Some extra guff in there of course to handle tabs going away.

I need to pore over that workaround-code a bit more to truly make sure it is doing the right thing. Even if it is though, it's jaaaaaaank. I need to [keep an eye on the following Chrome releases](https://bugs.chromium.org/p/chromium/issues/detail?id=1152255) (also cf [this issue](https://bugs.chromium.org/p/chromium/issues/detail?id=1305369) and possibly [this issue](https://bugs.chromium.org/p/chromium/issues/detail?id=1371876)) to see if this is addressed (either to break my workaround or make it unnecessary).

**Comment by neogeographica** [comment]
Timeout issue may be resolved in Chrome 110.0.5469.0.

https://bugs.chromium.org/p/chromium/issues/detail?id=1371876#c5

Looks like Dev channel has passed this point? Will try removing the timeout workaround when I get a chance.

edit: looking good...

**Comment by neogeographica** [comment]
Now it's broken in perhaps some different way in 111.0.5532.2. When the tab is opened on the "virtual URL" it fails to download anything with ERR_FILE_NOT_FOUND. Probably will need to do some printf-debugging.

**Comment by neogeographica** [comment]
The menu-item listener handler populates the blob, opens the tab on the "virtual URL", and exits. Then the fetch handler is just not firing at all. No errors in console.

chrome://serviceworker-internals/ does say "Fetch handler existence: EXISTS" for this service worker, so ... ?

Don't immediately see any relevant changes in the Chrome commits (but there's like a bazillion of those).

I've noticed that if I set the "capture" option to true when registering the listener -- documented as "indicating that events of this type will be dispatched to the registered listener before being dispatched to any EventTarget beneath it in the DOM tree" -- then things seem to work again? So maybe something else was intercepting and swallowing the virtual-URL fetch before my handler could see it?

I'll run it this way for a while as a test.
